### PR TITLE
Refactor Data Generators & Add flow from dataframe

### PR DESCRIPTION
**Pull request by jsalbert**
Refactor [data_generators.py](https://github.com/triagemd/keras-trainer/blob/add_flow_from_dataframe/keras_trainer/data_generators.py) to include just one class `EnhancedImageDataGenerator` 
New additions are the ability of passing dataframes to the `Trainer` class and the `class_mode` = 'probabilistic' parameter for the `flow_from_dataframe` [function](https://github.com/triagemd/keras-trainer/blob/add_flow_from_dataframe/keras_trainer/data_generators.py#L439).

### CODE REVIEW

File: keras_trainer/trainer.py [hunk 1]
```diff
@@ -107,15 +118,41 @@ def __init__(self, **options):
...
+                    batch_size=self.batch_size,
+                    x_col="filename",
+                    y_col="class_probabilities",
+                    iterator_mode=self.iterator_mode,
+                    n_outputs=self.n_outputs,
+                    target_size=self.model_spec.target_size[:2],
+                    class_mode='probabilistic'
+                )
+            else:
+                raise ValueError('`train_dataset_dir` must be specified')
```
Comment (adriaromero) [inline keras_trainer/trainer.py]: Either `train_dataset_dir ` or `train_dataset_dataframe`, right? 
```suggestion
                raise ValueError('`train_dataset_dir` or `train_dataset_dataframe` must be specified.')
```

File: keras_trainer/trainer.py [hunk 2]
```diff
@@ -107,15 +118,41 @@ def __init__(self, **options):
...
-            self.train_gen = self.train_data_generator.flow_from_directory(
-                self.train_dataset_dir,
-                batch_size=self.batch_size,
-                target_size=self.model_spec.target_size[:2],
-                class_mode='categorical'
-            )
-            self.num_classes = self.num_classes or self.train_gen.num_classes
-        else:
-            self.train_gen = self.train_generator
+            if self.train_dataset_dataframe is None and os.path.isdir(self.train_dataset_dir):
```
Comment (adriaromero) [inline keras_trainer/trainer.py]: Right now, if both `train_dataset_dir` and `train_dataset_dataframe` are provided, we're prioritizing the `train_dataset_dir` value. Is that what we want? Or we should raise an error?

File: tests/conftest.py [hunk 3]
```diff
@@ -8,11 +8,21 @@ def train_catdog_dataset_path():
     return os.path.abspath(os.path.join('tests', 'files', 'catdog', 'train'))
 
 
+@pytest.fixture('session')
+def train_catdog_dataframe_path():
```
[the remark this answers is not in the document]
Comment (jsalbert) [reply tests/conftest.py]: I thought about that but I didn't want to pass that many parameters to the test function.

[the remark this answers is not in the document]
Comment (jsalbert) [reply tests/conftest.py]: It is not the manifest but a dataframe which is the output of processing a manifest

File: keras_trainer/trainer.py [hunk 4]
```diff
@@ -124,34 +161,58 @@ def __init__(self, **options):
...
+                    directory=self.val_dataset_dir,
+                    batch_size=self.batch_size,
+                    x_col="filename",
+                    y_col="class_probabilities",
+                    n_outputs=self.n_outputs,
+                    target_size=self.model_spec.target_size[:2],
+                    class_mode='probabilistic'
+                )
+            else:
+                raise ValueError('`val_dataset_dir` must be specified')
```
[the remark this answers is not in the document]
Comment (jsalbert) [reply keras_trainer/trainer.py]: No, to run `flow_from_dataframe` you need the directory where the images are

[the remark this answers is not in the document]
Comment (jsalbert) [reply keras_trainer/trainer.py]: Unless you specify complete paths in the dataframe which we are not doing

File: keras_trainer/trainer.py [hunk 5]
```diff
@@ -46,17 +48,20 @@ class Trainer(object):
...
         'pooling': {'type': str, 'default': 'avg'},
+        'random_crop_size': {'type': float, 'default': None},
         'regularization_function': {'type': None, 'default': None},
         'regularization_layers': {'type': None, 'default': None},
         'regularize_bias': {'type': bool, 'default': False},
         'save_training_options': {'type': bool, 'default': True},
         'sgd_lr': {'type': float, 'default': 0.01},
         'top_layers': {'type': None, 'default': None},
-        'train_dataset_dir': {'type': str, 'default': None},
+        'track_sensitivity': {'type': bool, 'default': False},
```
[the remark this answers is not in the document]
Comment (jsalbert) [reply keras_trainer/trainer.py]: I'll keep it as it is right now

File: keras_trainer/trainer.py (hunk 2 above)
Comment (jsalbert) [reply keras_trainer/trainer.py]: `train_dataset_dir` is always required, we are prioritizing `train_dataset_dataframe` if it exists

File: tests/files/catdog/train_data.json [hunk 6]
```diff
@@ -0,0 +1,32 @@
+[
+  {
+    "filename": "cat/cat-1.jpg",
+    "class_probabilities": [0.9, 0.1],
```
Comment (jeremykawahara) [inline tests/files/catdog/train_data.json]: so the idea is to store a list of probabilities for all the potential classes here, and then have another list that has the relative ordering of the class names (e.g., index 0 = cat, index 1 = dog).

I'm wondering if we should consider that likely have sparse labels (i.e., we'll have a large list of potential classes, with most classes getting a 0 probability).

Is there a case to be made that we should only store the non-zero class labels? Something like,

```
"class_probabilities": {
   "TRIAGE:CAT": 0.9,
   "TRIAGE:DOG":0.1,
}
```

File: keras_trainer/trainer.py [hunk 7]
```diff
@@ -91,10 +96,16 @@ def __init__(self, **options):
             'options': options
         }
 
+        if isinstance(self.loss_function, list) and len(self.loss_function) > 1:
+            self.n_outputs = len(self.loss_function)
```
[the remark this answers is not in the document]
Comment (jsalbert) [reply keras_trainer/trainer.py]: The number of output labels that we will generate. 
The number of model outputs is defined by a custom model.

File: keras_trainer/trainer.py [hunk 8]
```diff
@@ -266,15 +327,21 @@ def run(self):
             loss_weights=self.loss_weights
         )
 
+        if self.train_generator.samples // self.batch_size == 0:
+            raise ValueError('Batch size is higher than the total number of training samples')
+
+        if self.val_generator.samples // self.batch_size == 0:
+            raise ValueError('Batch size is higher than the total number of validation samples')
```
[the remark this answers is not in the document]
Comment (jsalbert) [reply keras_trainer/trainer.py]: I don't think the Trainer class and run method works for no validation set at the moment. If we wanted to do that maybe it would be a good addition. I added the check because I was getting errors and I believe now it is easier to debug.

[the remark this answers is not in the document]
Comment (jeremykawahara) [reply keras_trainer/trainer.py]: cool ya makes sense! Looks good

File: tests/files/catdog/train_data.json (hunk 6 above)
Comment (jsalbert) [reply tests/files/catdog/train_data.json]: What you said would need that extra list with class names and indices. 
It could be added in the future, as the [code](https://github.com/triagemd/ontology/blob/master/ontology/ontology.py#L768) in ontology can give us both formats. The one I used was the easiest one, because it required less processing from my side.

File: keras_trainer/trainer.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (jeremykawahara) [reply keras_trainer/trainer.py]: is the "number of output labels" the same as the number of classes? I don't see the relationship between the number of loss functions and the number of output labels?

File: tests/files/catdog/train_data.json (hunk 6 above)
Comment (jsalbert) [reply tests/files/catdog/train_data.json]: Also with that approach we would need to store much more columns, which I am not sure if it would be an issue.

File: keras_trainer/trainer.py (hunk 2 above)
Comment (jsalbert) [reply keras_trainer/trainer.py]: True! Thanks for the simplification 👀

File: keras_trainer/trainer.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (jsalbert) [reply keras_trainer/trainer.py]: Means the number of times you have to return `batch_y` to be able to run it in  the keras `fit_generator` function. 
We do `[batch_y] * n_outputs` where n_outputs is the len(loss_functions). Because they require one label batch per output.

[the remark this answers is not in the document]
Comment (jsalbert) [reply keras_trainer/trainer.py]: There is an example in the notebook. I added that parameter to handle it internally. It will only work if your loss functions take the same output labels though.

File: keras_trainer/data_generators.py [hunk 9]
```diff
@@ -474,17 +194,54 @@ def _get_batches_of_transformed_samples(self, index_array):
...
+                # Class array divides the 1D array into an index list per each class.
+                # E.g. [0, 0, 1, 1] --> [array([0, 1]), array([3, 4)]
+                self.num_classes = len(np.unique(classes))
+                self.class_array = [np.where(i == classes)[0] for i in range(0, self.num_classes)]
+            else:
+                raise ValueError('`classes` variable cannot be None')
+
+    def _set_index_array(self):
+        if self.mode is None:
+            self.index_array = np.arange(self.n)
```
[the remark this answers is not in the document]
Comment (jsalbert) [reply keras_trainer/data_generators.py]: Number of samples
